File: Algorithms/C++/00014-Longest Common Prefix.cpp
class Solution {
public:
    string longestCommonPrefix(vector<string>& strs) {
        std::string ans;
        std::string first = strs[0];
        int shortest{INT_MAX};

        for(int i{0}; i < strs.size(); i++){
            int strSize = strs[i].size();
            shortest = std::min(shortest, strSize);
        }

        for(int i{0}; i < shortest; i++){
            char prefixChar = strs[0][i];    // ith char of first string
            for(int j{0}; j < strs.size(); j++){
                // found a string that breaks prefixChar condition
                if(strs[j][i] != prefixChar){
                    return ans;
                }
            }    
            // finished one iteration of all strings
            // add prefixChar to ans string
            ans += prefixChar;
        }

        return ans;
    }
};

/*
    O(n) where n = total number of characters in all strings
    O(1)

    - iterate through every character of every string
    - if prefix does not match, immediately return ans


*/

File: Algorithms/Python3/00014-Longest Common Prefix.py
class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        strs = sorted(strs, key=len)
        result = ""
        for i, c in enumerate(strs[0]):
            for word in range(len(strs)):
                if strs[word][i] != c:
                    return result
            result += c 

        return result


'''
    sort strs by shortest word first OR swap shortest word to first index

    take the first character of the first string
    iterate through each word in strs to see if each first char matches that char
    if yes, then append char to our output string
    if no, then return output string
    repeat until no more characters

    tc: O(nlogn)
    sc: O(n)


'''